packages/analytics-react-native/test/react-native-sessions.test.ts

```diff
@@ -0,0 +1,523 @@
+    expect(events[2].session_id).toEqual(950);
+    expect(events[2].time).toEqual(1501);
+
+    expect(events[3].event_type).toEqual('session_start');
+    expect(events[3].session_id).toEqual(3000);
+    expect(events[3].time).toEqual(3000);
+
+    expect(events[4].event_type).toEqual('event-2');
+    expect(events[4].session_id).toEqual(3000);
+    expect(events[4].time).toEqual(3000);
+  });
+
+  test('session data should be persisted', async () => {
+    const send = jest.fn().mockReturnValue(sendResponse);
+    const cookieStorage = new core.MemoryStorage<UserSession>();
+
+    const client1 = new AmplitudeReactNativeTest(950);
+    await client1.init(API_KEY, undefined, clientOptions(send, cookieStorage, true)).promise;
+
+    client1.setForeground(1000);
+
+    expect(client1.getSessionId()).toEqual(950);
+    expect(client1.config.sessionId).toEqual(950);
+    expect(client1.config.lastEventTime).toEqual(1000);
+    expect(client1.config.lastEventId).toEqual(1);
+
+    client1.track(createEvent(1200, 'event-1'));
+
+    expect(client1.getSessionId()).toEqual(950);
+    expect(client1.config.sessionId).toEqual(950);
+    expect(client1.config.lastEventTime).toEqual(1200);
+    expect(client1.config.lastEventId).toEqual(2);
+
+    const client2 = new AmplitudeReactNativeTest(1250);
+    await client2.init(API_KEY, undefined, clientOptions(send, cookieStorage, true)).promise;
+
+    expect(client2.getSessionId()).toEqual(950);
+    expect(client2.config.sessionId).toEqual(950);
+    expect(client2.config.lastEventTime).toEqual(1250);
+    expect(client1.config.lastEventId).toEqual(2);
+  });
+
+  test('explicit session for event should be preserved', async () => {
+    const send = jest.fn().mockReturnValue(sendResponse);
+    const cookieStorage = new core.MemoryStorage<UserSession>();
+
+    const client = new AmplitudeReactNativeTest(950);
+    await client.init(API_KEY, undefined, clientOptions(send, cookieStorage, true)).promise;
+
+    client.track(createEvent(1000, 'event-1'));
+    client.track(createEvent(1050, 'event-2', 3000));
+    client.track(createEvent(1100, 'event-3'));
+    await client.flush().promise;
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    const events = send.mock.calls.flatMap((call) => call[1].events as Event[]);
+    expect(events.length).toEqual(4);
+
+    expect(events[0].event_type).toEqual('session_start');
+    expect(events[0].session_id).toEqual(950);
+    expect(events[0].time).toEqual(950);
+
+    expect(events[1].event_type).toEqual('event-1');
+    expect(events[1].session_id).toEqual(950);
+    expect(events[1].time).toEqual(1000);
+
+    expect(events[2].event_type).toEqual('event-2');
+    expect(events[2].session_id).toEqual(3000);
+    expect(events[2].time).toEqual(1050);
+
+    expect(events[3].event_type).toEqual('event-3');
+    expect(events[3].session_id).toEqual(950);
+    expect(events[3].time).toEqual(1100);
```

Comment:
Not 100% what should happen here. Some options

1. 1 session like you have now
2. "1.5" sessions - see code below.
3. 2.5 sessions - same as 2 but with session_start/end events around event-2.

I think 3 would be too hard to implement, but what are your thoughts on 2? Should an out of session event extend the current session?

```
'session_start', 950, 950
'event-1', 950, 1000
'session_end', 950, 1001

'event-2', 3000, 1050

'session_start', 1100, 1099
'event-3', 1100, 1100
```

Reply:
@justin-fiedler 
Maybe the test is not realistic. Based on documentation (https://www.docs.developers.amplitude.com/data/sdks/browser-2/#custom-session-id) ` When setting a custom session ID, make sure the value is in milliseconds since epoch` - I violated it in the test (seesion _id 3000 is greater than timestamp 1050).
I would stay with 1 for now - I think, the behavior should be the same in all client SDKs. WDYT?

Reply:
I discussed with the team and we feel that option #2 is the right approach, i.e. out-of-session should not extend the current session.

In general, if the `event.sessionId` doesn't equal `config.sessionId` then it shouldn't update `lastEventTime`

Reply:
@justin-fiedler Please, take a look at updated code and tests